src/Microsoft.OData.Client/DataServiceContext.cs

```diff
@@ -3935,13 +4018,89 @@ private void DeleteObjectInternal(object entity, bool failIfInAddedState)
                 this.entityTracker.IncrementChange(resource);
             }
         }
 
+        /// <summary>
+        /// Mark an existing object being tracked by the context for deletion.
+        /// </summary>
+        /// <param name="entity">entity to be mark deleted</param>
+        /// <param name="failIfInAddedState">If true, then an exception will be thrown if the entity is in the added state.</param>
+        /// <exception cref="ArgumentNullException">if entity is null</exception>
+        /// <exception cref="InvalidOperationException">if entity is not being tracked by the context, or if the entity is in the added state and <paramref name="failIfInAddedState"/> is true.</exception>
+        /// <remarks>
+        /// Existing objects in the Added state become detached if <paramref name="failIfInAddedState"/> is false.
+        /// </remarks>
+        private void DeleteObjectInternal(object entity, bool failIfInAddedState)
+        {
+            this.DeleteObjectInternal(entity, failIfInAddedState, null);
+        }
+
+        private static void GetDependsOnChangeOrdersAndChangeSetIds(
+            List<string> dependsOnIds,
+            IEnumerable<EntityDescriptor> entities,
+            ClientEdmModel model,
+            out List<string>  dependsOnIdsAsChangeOrders,
+            out HashSet<string> dependsOnChangeSetIds)
+        {
+            dependsOnIdsAsChangeOrders = new List<string>();
+            dependsOnChangeSetIds = new HashSet<string>();
+
+            foreach (string dependOnId in dependsOnIds)
+            {
+                foreach (EntityDescriptor entityDescriptor in entities)
+                {
+                    object entity = (object)entityDescriptor.Entity;
+                    string entityName = entity.GetType().FullName;
+                    string key = GetKeyPropertyName(model, entityName);
+                    object id = entity.GetType().GetProperty(key).GetValue(entity, null);
+                    if (dependOnId == id.ToString())
+                    {
+                        uint changeOrder = entityDescriptor.ChangeOrder;
+                        dependsOnIdsAsChangeOrders.Add(changeOrder.ToString(CultureInfo.InvariantCulture));
+                        dependsOnChangeSetIds.Add(entityDescriptor.ChangeSetId);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the name of the Declared Key for an entity.
+        /// </summary>
+        /// <param name="model"></param>
+        /// <param name="entityName"></param>
+        /// <returns>Declared key property name.</returns>
+        private static string GetKeyPropertyName(ClientEdmModel model, string entityName)
+        {
+            EdmEntityTypeWithDelayLoadedProperties schemaType = (EdmEntityTypeWithDelayLoadedProperties)model.FindDeclaredType(entityName);
+            return schemaType.DeclaredKey.First().Name;

```

Comment:
@KenitoInc If we only pick the first property, would the `dependOn` feature be foolproof when the entity has a compound key?

Reply:
See this comment https://github.com/OData/odata.net/pull/1806#discussion_r610376573